From 850c82b7c4a144cec6e56cff515967474b80920d Mon Sep 17 00:00:00 2001
From: Renoir Boulanger <renoir@w3.org>
Date: Thu, 29 Jan 2015 16:07:32 -0500
Subject: [PATCH] =?UTF-8?q?Block=20profile=20update=20to=20users=20who=20d?=
 =?UTF-8?q?idn=E2=80=99t=20confirm=20their=20email?=
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Use already define MW variable $wgEmailConfirmToEdit and enforces
users to have a confirmed email address to edit their profile.

To ensure a user has a confirmed email, I used $user->getEmailAuthenticationTimestamp().
If the return value is NULL, it means the user didn’t click the confirm link.

Change-Id: I0dc8c36f77360b5fb437102317349e0be1d1cb36
---
 UserProfile/SpecialUpdateProfile.php | 11 ++++++++++-
 UserProfile/i18n/en.json             |  1 +
 UserProfile/i18n/fr.json             |  1 +
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/UserProfile/SpecialUpdateProfile.php b/UserProfile/SpecialUpdateProfile.php
index 87124b1..ffb2fe8 100644
--- a/UserProfile/SpecialUpdateProfile.php
+++ b/UserProfile/SpecialUpdateProfile.php
@@ -51,7 +51,7 @@ class SpecialUpdateProfile extends UnlistedSpecialPage {
 	 * @param $section Mixed: parameter passed to the page or null
 	 */
 	public function execute( $section ) {
-		global $wgUpdateProfileInRecentChanges, $wgUserProfileThresholds, $wgSupressPageTitle, $wgAutoConfirmCount;
+		global $wgUpdateProfileInRecentChanges, $wgUserProfileThresholds, $wgSupressPageTitle, $wgAutoConfirmCount, $wgEmailConfirmToEdit;
 
 		$out = $this->getOutput();
 		$request = $this->getRequest();
@@ -109,6 +109,11 @@ class SpecialUpdateProfile extends UnlistedSpecialPage {
 			$hasEqualEditThreshold = ( isset( $wgUserProfileThresholds['edit'] ) && $wgUserProfileThresholds['edit'] == $wgAutoConfirmCount ) ? true : false;
 			$can_create = ( $user->isAllowed( 'createpage' ) && $hasEqualEditThreshold ) ? true : $can_create;
 
+			// Ensure we enforce profile creation exclusively to members who confirmed their email
+			if ( $user->getEmailAuthenticationTimestamp() === NULL && $wgEmailConfirmToEdit === true ) {
+				$can_create = false;
+			}
+
 			// Boo, go away!
 			if ( $can_create == false ) {
 				global $wgSupressPageTitle;
@@ -140,6 +145,10 @@ class SpecialUpdateProfile extends UnlistedSpecialPage {
 					*/
 					$thresholdMessages[] = $this->msg( 'user-profile-create-threshold-' . $reason )->numParams( $requiredAmount )->parse();
 				}
+				// Set a useful message of why.
+				if ( $user->getEmailAuthenticationTimestamp() === NULL && $wgEmailConfirmToEdit === true ) {
+						$thresholdMessages[] = $this->msg( 'user-profile-create-threshold-only-confirmed-email' )->text();
+				}
 				$out->addHTML(
 					$this->msg( 'user-profile-create-threshold-reason',
 						$this->getLanguage()->commaList( $thresholdMessages )
diff --git a/UserProfile/i18n/en.json b/UserProfile/i18n/en.json
index e7295a0..e545655 100644
--- a/UserProfile/i18n/en.json
+++ b/UserProfile/i18n/en.json
@@ -163,6 +163,7 @@
 	"user-profile-goback": "Go back",
 	"user-profile-userlevels-link": "User levels",
 	"user-profile-create-threshold-title": "Create user profile",
+	"user-profile-create-threshold-only-confirmed-email": "and a confirmed email address",
 	"user-profile-create-threshold-edits": "{{PLURAL:$1|one edit|$1 edits}}",
 	"user-profile-create-threshold-votes": "{{PLURAL:$1|one vote|$1 votes}}",
 	"user-profile-create-threshold-comments": "{{PLURAL:$1|one comment|$1 comments}}",
diff --git a/UserProfile/i18n/fr.json b/UserProfile/i18n/fr.json
index e8f9a5a..8a163cc 100644
--- a/UserProfile/i18n/fr.json
+++ b/UserProfile/i18n/fr.json
@@ -193,5 +193,6 @@
 	"user-profile-picture-log-entry": "a téléversé un nouvel avatar",
 	"user-profile-picture-log-delete-entry": "a supprimé l’avatar de $1",
 	"right-avatarremove": "Supprimer les avatars d’un utilisateur",
+	"user-profile-create-threshold-only-confirmed-email": "et une addresse courriel confirmée",
 	"right-editothersprofiles": "Mettre à jour le profil social d'autres utilisateurs"
 }
-- 
1.9.3 (Apple Git-50)

